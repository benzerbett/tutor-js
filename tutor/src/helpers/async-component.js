import { asyncComponent as asyncLoader } from 'react-async-component';
import ErrorComponent from '../components/error-monitoring/async-load-error';
import LoadingComponent from '../components/loading-screen';

export function asyncComponent(resolve) {
  return asyncLoader({
    resolve,
    LoadingComponent,
    ErrorComponent,
  });
}

export function loadAsync(resolve) {
  // return a function so the router will only evaluate it when it's needed
  // in the future we can insert role dependant logic here
  return () => asyncComponent(resolve);
}
