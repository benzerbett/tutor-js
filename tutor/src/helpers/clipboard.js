import _ from 'underscore';

const SUPPORTED = typeof document.queryCommandSupported === 'function' ? document.queryCommandSupported('copy') : undefined;

export default {
  isSupported: _.constant(!!SUPPORTED),

  copy() {
    if (!SUPPORTED) { return; }

    try {
      return document.execCommand('copy');
    } catch (e) {
      return console.warn('clipboard copy failed', e);
    }
  },
};
