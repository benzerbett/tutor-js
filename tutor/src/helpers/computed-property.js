import { forIn } from 'lodash';
import { computed } from 'mobx';

export function addComputedProperty(obj, name, fn) {

  Object.defineProperty(obj, name, {
    configurable: true,
    get() {
      const getter = computed(fn, { context: this });
      Object.defineProperty(this, name, {
        configurable: false,
        get: () => getter.get(),
      });
      return getter.get();
    },
  });

}

export function extendHasMany(properties) {
  return (collection) => {
    forIn(properties, (fn, prop) => {
      addComputedProperty(collection, prop, fn);
    });
  };
}
