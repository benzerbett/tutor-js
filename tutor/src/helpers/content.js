import _ from 'underscore';


export default {
  // Convert each number to base 10 with it's position based on index.
  // If section is not present, 0 is set for it
  //   1 becomes 100, 1.1 becomes 101, 3.2 to 302, 3.2.1 -> 30201
  // Useful for sorting
  chapterSectionToNumber(chapter_section) {
    if (_.isString(chapter_section)) { chapter_section = chapter_section.split('.'); }
    if (chapter_section.length === 1) { chapter_section.push(0); } // add a section 0 if it has only a chapter
    let position = -1;
    return _.reduceRight(chapter_section, (memo, num) => memo + (num * Math.pow(100, (position += 1)))
      , 0);
  },
};
