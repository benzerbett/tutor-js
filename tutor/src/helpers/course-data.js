import Courses from '../models/courses-map';

import Router from '../helpers/router';
import TimeHelper from '../helpers/time';

export default {
  getCourseDataProps(courseId) {
    let dataProps;
    if (courseId == null) {
      ((((((({ courseId } = Router.currentParams())))))));
    }
    const course = Courses.get(courseId);
    return dataProps = {
      'data-title': course.name,
      'data-book-title': course.bookName || '',
      'data-appearance': course.appearance_code,
    };
  },

  getCourseBounds(courseId) {
    const course = Courses.get(courseId);

    const start = TimeHelper.getMomentPreserveDate(course.starts_at, TimeHelper.ISO_DATE_FORMAT);
    const end = TimeHelper.getMomentPreserveDate(course.ends_at, TimeHelper.ISO_DATE_FORMAT);

    return { start, end };
  },
};
