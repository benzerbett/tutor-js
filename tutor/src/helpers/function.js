export default {
  // coffeelint: disable=no_empty_functions
  empty() {},
};
// coffeelint: enable=no_empty_functions
