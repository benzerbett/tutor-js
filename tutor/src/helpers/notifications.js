import { NotificationActions } from 'shared';

export default {
  start(bootstrapData) {
    NotificationActions.startPolling();
    return (() => {
      const result = [];
      const object = bootstrapData.flash || {};
      for (let level in object) {
        const message = object[level];
        result.push(NotificationActions.display({ message, level }));
      }
      return result;
    })();
  },

  buildCallbackHandlers(comp) {
    // Require router here because requiring it in global scope
    // causes hot module reloading to fail in dev mode because this file is required outside the HMR scope
    const Router = require('./router');
    const { router } = comp.context;
    if (!router) {
      throw new Error('Component\'s context must have router present');
    }
    return {
      missing_student_id: {
        onAdd({ course }) {
          return router.history.push(
            Router.makePathname('changeStudentId', { courseId: course.id })
          );
        },
      },
    };
  },
};
