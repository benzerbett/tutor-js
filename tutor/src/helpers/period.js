import { filter, isEmpty, sortBy, toNumber, isNumber } from 'lodash';
import { addComputedProperty } from './computed-property';
import S from './string';

// Used to filter periods by helper methods
const isArchivedCheckFn = period => period.is_archived;
const isActiveCheckFn   = period => !period.is_archived;

const NumberLike = /[^0-9]+/;


const PeriodHelper = {
  getOrder(period) {
    return S.getNumberAndStringOrder(period.name);
  },

  sort(periods) {
    // expects either numbers, names with numbers or just names
    periods = sortBy(periods, (period) => { // eslint-disable-line consistent-return
      if (!isNumber(period.name)) {
        const name = period.name.match(/[^0-9]+/ig);
        if (name) {
          return name;
        }
      }
    });
    return sortBy(periods, (period) => { // eslint-disable-line consistent-return
      if (!isNumber(period.name)) {
        const number = period.name.match(/[0-9.-]+/g);
        if (number) {
          return parseFloat(number);
        }
      } else {
        return period.name;
      }
    });
  },

};


export default PeriodHelper;
