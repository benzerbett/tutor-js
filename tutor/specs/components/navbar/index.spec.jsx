import Navbar from '../../../src/components/navbar';
import { Router, Provider } from '../../helpers';
import TourContext from '../../../src/models/tour/context';

jest.mock('../../../src/models/chat');

describe('Main Navbar', () => {
  it('renders and matches snapshot', () => {
    const tourContext = new TourContext();
    expect.snapshot(
      <Router>
        <Provider tourContext={tourContext}>
          <Navbar.bar />
        </Provider>
      </Router>
    ).toMatchSnapshot();
  });
});
