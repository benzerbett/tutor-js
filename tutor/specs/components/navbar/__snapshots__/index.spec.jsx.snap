// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main navbar renders and matches snapshot 1`] = `
<nav
  className="tutor-top-navbar"
>
  <div
    className="tutor-nav-controls"
  >
    <div
      className="left-side-controls"
    >
      <a
        aria-label="dashboard"
        className="brand"
        href="/dashboard"
        onClick={[Function]}
      >
        <i
          className="ui-brand-logo"
        />
      </a>
    </div>
    <div
      className="right-side-controls"
    >
      <div
        className="support-menu dropdown btn-group"
      >
        <a
          aria-expanded={false}
          aria-haspopup={true}
          aria-label="Page tips and support"
          className="dropdown-toggle"
          href="#"
          id="support-menu"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
        >
          <div
            aria-labelledby="support-menu"
            className="tour-anchor"
            data-tour-anchor-id="support-menu-button"
            onSelect={[Function]}
          >
            <i
              className="tutor-icon fa fa-question-circle"
              role="presentation"
              type="question-circle"
            />
            <span
              className="control-label"
              title="Page tips and support"
            >
              Help
            </span>
            <i
              className="tutor-icon fa fa-angle-down toggle"
              role="presentation"
              type="angle-down"
            />
          </div>
        </a>
        <ul
          aria-labelledby="support-menu"
          className="dropdown-menu"
          role="menu"
        >
          <li
            className="-help-link"
            role="presentation"
            style={undefined}
          >
            <a
              href="https://openstax.secure.force.com/help"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1"
              target="_blank"
            >
              <span>
                Help Articles
              </span>
            </a>
          </li>
          <li
            className="support-document-link"
            role="presentation"
            style={undefined}
          >
            <a
              href="https://d3bxy9euw4e147.cloudfront.net/oscms-prodcms/media/documents/openstax_tutor_student_getting_started_guide.pdf"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1"
              target="_blank"
            >
              <div
                className="tour-anchor"
                data-tour-anchor-id="menu-support-document"
              >
                Getting Started Guide
              </div>
            </a>
          </li>
          <li
            className="-help-link"
            role="presentation"
            style={undefined}
          >
            <a
              href="/accessibility-statement/"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1"
            >
              <span>
                Accessibility Statement
              </span>
            </a>
          </li>
        </ul>
      </div>
      <div
        className="user-menu dropdown btn-group"
      >
        <a
          aria-expanded={false}
          aria-haspopup={true}
          aria-label="Account settings"
          className="dropdown-toggle"
          href="#"
          id="user-menu"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
        >
          <span
            className="initials"
          >
            
          </span>
        </a>
        <ul
          aria-labelledby="user-menu"
          className="dropdown-menu dropdown-menu-right"
          role="menu"
        >
          <li
            className="logout"
            role="presentation"
            style={undefined}
          >
            <a
              href="#"
              label="Log out"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1"
            >
              <form
                acceptCharset="UTF-8"
                action="/accounts/logout"
                className="-logout-form"
                method="post"
              >
                <input
                  name="_method"
                  type="hidden"
                  value="delete"
                />
                <input
                  name="authenticity_token"
                  type="hidden"
                  value={undefined}
                />
                <input
                  aria-label="Log out"
                  type="submit"
                  value="Log out"
                />
              </form>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</nav>
`;
